src/abstractdataframe/show.jl

```diff
@@ -573,61 +573,213 @@ function _show(io::IO,
                df::AbstractDataFrame;
                allrows::Bool = !get(io, :limit, false),
                allcols::Bool = !get(io, :limit, false),
-               splitcols = get(io, :limit, false),
                rowlabel::Symbol = :Row,
                summary::Bool = true,
                eltypes::Bool = true,
-               rowid=nothing,
-               truncstring::Int)
+               rowid = nothing,
+               truncate::Int = 32,
+               kwargs...)
+
     _check_consistency(df)
 
-    # we will pass around this buffer to avoid its reallocation in ourstrwidth
-    buffer = IOBuffer(Vector{UInt8}(undef, 80), read=true, write=true)
+    aux = names(df)
+    names_len = textwidth.(aux)
+    maxwidth = max.(9, names_len)
+    names_mat = permutedims(aux)
+    types = eltype.(eachcol(df))
+
+    # NOTE: If we reuse `types` here, the time to print the first table is 2x more.
+    # This should be something related to type inference.
+    types_str = permutedims(compacttype.(eltype.(eachcol(df)), maxwidth))
+
+    if allcols && allrows
+        crop = :none
+    elseif allcols
+        crop = :vertical
+    elseif allrows
+        crop = :horizontal
+    else
+        crop = :both
+    end
 
-    nrows = size(df, 1)
-    if rowid !== nothing
-        if size(df, 2) == 0
-            rowid = nothing
-        elseif nrows != 1
-            throw(ArgumentError("rowid may be passed only with a single row data frame"))
+    compact_printing::Bool = get(io, :compact, true)
+
+    num_rows, num_cols = size(df)
+
+    # By default, we align the columns to the left unless they are numbers,
+    # which is checked in the following.
+    alignment = fill(:l, num_cols)
+
+    # This vector stores the column indices that are only floats. In this case,
+    # the printed numbers will be aligned on the decimal point.
+    float_cols = Int[]
+
+    # These vectors contain the number of the row and the padding that must be
+    # applied so that the float number is aligned with on the decimal point
+    indices = Vector{Int}[]
+    padding = Vector{Int}[]
```

Comment:
No hurry to change this now, but I wonder whether it wouldn't be faster to avoid storing this information. Instead, the padding can be computed very cheaply in `_pretty_tables_float_formatter` once you have obtained the string representation of the number using `show`: it's just the difference between the string length and the maximum size (since numbers are pure ASCII it's even the number of codeunits, which is super fast to compute).

(Actually I must say I don't completely get why this algorithm is slow. Yes, it requires going over the data twice, but usually printing is much slower than this step which only involves quite fast operations. Isn't the slowness due to the fact that `_pretty_tables_float_formatter` is quite inefficient as it checks the row and column index for each row?)

Reply:
> No hurry to change this now, but I wonder whether it wouldn't be faster to avoid storing this information. Instead, the padding can be computed very cheaply in `_pretty_tables_float_formatter` once you have obtained the string representation of the number using `show`: it's just the difference between the string length and the maximum size (since numbers are pure ASCII it's even the number of codeunits, which is super fast to compute).
> 
> (Actually I must say I don't completely get why this algorithm is slow. Yes, it requires going over the data twice, but usually printing is much slower than this step which only involves quite fast operations. Isn't the slowness due to the fact that `_pretty_tables_float_formatter` is quite inefficient as it checks the row and column index for each row?)

I am not entirely certain that I understood it correctly. Let's say you have just two rows: 100.0 and 0.001. The maximum size would be 5 in both. Hence, the padding in this case obtained by the algorithm you proposed would be 0. However, the padding in the first case must be 0 and in the second case must be 2. Am I missing something?

> (Actually I must say I don't completely get why this algorithm is slow. Yes, it requires going over the data twice, but usually printing is much slower than this step which only involves quite fast operations. Isn't the slowness due to the fact that `_pretty_tables_float_formatter` is quite inefficient as it checks the row and column index for each row?)

Yes, it can be! Do you have any suggestions to improve this?

Reply:
> I am not entirely certain that I understood it correctly. Let's say you have just two rows: 100.0 and 0.001. The maximum size would be 5 in both. Hence, the padding in this case obtained by the algorithm you proposed would be 0. However, the padding in the first case must be 0 and in the second case must be 2. Am I missing something?

Woops, indeed that doesn't work. You'd need to recompute the order. Tough that should be quite fast so maybe it's not worse than allocating a full array.


> Yes, it can be! Do you have any suggestions to improve this?

Without changing the current design based on passing a function to PrettyTables, it seems to me that the main way to make things faster would be to avoid having to do a search of the row and column index for each cell. It should be much faster to store Boolean vectors with one entry for each row/column and index into these with `i` and `j`. Though for rows that means storing a quite large vector; another approach would be to pass the ranges of rows as two separate arguments (checking for inclusion in a range is optimized to simple comparisons).

It could also be faster to call `show(io, ...)` directly on the output stream rather than using `sprint` to get a string. But I'm not sure whether it's possible with PrettyTables.